CalendarProject/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -0,0 +1,128 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+
+public final class FindMeetingQuery {
+    /* 
+    * Given  alist of events in one day and a request for a meeting, 
+    * returns all the timeRanges when the meetik can take place so that each attendee can attend without having anothe roverlapping event.
+    */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        events = new ArrayList(events);
+        events.removeIf(event -> !existCommonAttendees(event.getAttendees(), request.getAttendees()));
+        List<TimeRange> unavailableTimes = getListOfTimeRanges(events);
+        List<TimeRange> reducedUnavailableTimes = getReducedListOfTimeRanges(unavailableTimes);
+        List<TimeRange> availableTimeRanges = getComplementerTimeRanges(reducedUnavailableTimes);
+        removeTooShortTimeRanges(availableTimeRanges, request.getDuration());
+        return availableTimeRanges;
+    }
+
+    /* 
+    * Removes from the list the time ranges which are shorter than minDuration.
+    */ 
+    private void removeTooShortTimeRanges(List<TimeRange> timeRanges, long minDuration) {
+        timeRanges.removeIf(timeRange -> timeRange.duration() < minDuration);
+    }
+
+    /* 
+    * Given the timeRanges in one day, this function returns the list of complementerTimeRanges that are in the same day, but are not present in timeRanges.
+    * The union of timeRanges and complementerTimeRanges is an entire day.
+    */
+    private List<TimeRange> getComplementerTimeRanges(List<TimeRange> timeRanges) {
+        List<TimeRange> complementerTimeRanges = new LinkedList<TimeRange>();
```

Comment:
Why LinkedList?

Reply:
Because I need to remove some elements later from the list(see removeTooShortTimeRanges()), and as far as I know, removals can be done in O(1) time complexity on a LinkedList, but only O(n) on an ArrayList. 
Moreover, no index-based access of element is required, which would make using an ArrayLists preferable.

Reply:
Good point.

Although I think Java has some optimization for removeIf so the performance is not much different between ArrayList and LinkedList.